### ISSUE TITLE
Auto Differentiation, complex gradient,  is it expected?

### ISSUE THREAD

**Issue by carlos-luque** [qiboteam/qiboml#185]
I have attempted to run a test based on the test_variable_backpropagation to verify if the gradient becomes complex when the tf.variable is of type complex128, as indicated by the comment on line 897."

I ran the code:

```py
from qibo.backends import construct_backend
backend = construct_backend("tensorflow")
import tensorflow as tf


theta = tf.Variable(0.1234, dtype=tf.complex128)

with tf.GradientTape() as tape:
    tape.watch(theta)
    c = Circuit(1)
    c.add(gates.X(0))
    c.add(gates.RZ(0, theta))
    result = backend.execute_circuit(c) 
    loss = tf.math.real(result.state()[-1])
    print ("state:")
    print(result.state())
grad = tape.gradient(loss, theta)
print("gradient:")
print(grad)
```

This produced:
```py
state:
tf.Tensor([0.        +0.j         0.99809716+0.06166086j], shape=(2,), dtype=complex128)
gradient
tf.Tensor((-0.030830429966003096-0.4990485793869284j), shape=(), dtype=complex128)
```
Does the appearance of a complex gradient imply a modification in the target function? Is this an anticipated outcome, or does it suggest a potential issue?
